File: Data Structure/Map/LeetCode/maximal-network-rank.cpp
class Solution {
public:
    int maximalNetworkRank(int n, vector<vector<int>>& roads) {
        vector<int> adj[n];
        map<pair<int, int>, int> mp;
        for(auto &it: roads){
            adj[it[0]].push_back(it[1]);
            adj[it[1]].push_back(it[0]);
            mp[{it[0],it[1]}] = 1;
            mp[{it[1], it[0]}] = 1;
        }
        int ans = 0;
        for(int i = 0; i < n; i++){
            for(int j = i+1; j < n; j++){
                int cal = adj[i].size() + adj[j].size();
                if(mp[{i, j}] == 1 || mp[{j, i}] == 1){
                    cal -= 1;
                }
                ans = max(ans, cal);
            }
        }
        return ans;
    }
};

File: Programming Language/Python3/maximal-network-rank.py
class Solution:
    def maximalNetworkRank(self, n: int, roads: List[List[int]]) -> int:
        adj = {}
        mp = {}
        
        for i in range(n):
            adj[i] = []

        for u, v in roads:
            adj[u].append(v)
            adj[v].append(u)
            mp[(u, v)] = 1
            mp[(v, u)] = 1

        ans = 0
        for i in range(n):
            for j in range(i+1, n):
                cal = len(adj[i]) + len(adj[j])
                if (i, j) in mp or (j, i) in mp:
                    cal -= 1
                ans = max(ans, cal)

        return ans
